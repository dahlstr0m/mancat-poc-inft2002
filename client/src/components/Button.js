// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a success button using Bootstrap styles.
 */
class ButtonSuccess extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-success' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a danger button using Bootstrap styles.
 */
class ButtonDanger extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-danger' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a light button using Bootstrap styles.
 */
class ButtonLight extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-light' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a button using Bootstrap styles.
 */
export default class Button {
  static Success = ButtonSuccess;
  static Danger = ButtonDanger;
  static Light = ButtonLight;
}
