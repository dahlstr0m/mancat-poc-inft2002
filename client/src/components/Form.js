// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a form label using Bootstrap styles.
 */
class FormLabel extends Component<{ children?: React.Node }> {
  render() {
    return <label className="col-form-label">{this.props.children}</label>;
  }
}

/**
 * Renders a form input using Bootstrap styles.
 */
class FormInput extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders a form textarea using Bootstrap styles.
 */
class FormTextarea extends React.Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLTextAreaElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, rows, cols
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, ...rest } = this.props;
    return <textarea {...rest} className="form-control" value={value} onChange={onChange} />;
  }
}

/**
 * Renders a form checkbox using Bootstrap styles.
 */
class FormCheckbox extends Component<{
  checked: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
  labeling: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { labeling, checked, onChange, ...rest } = this.props;
    return (
      <div className="custom-control custom-checkbox">
        <input
          {...rest}
          type="checkbox"
          className="custom-control-input"
          id="customCheck"
          checked={checked}
          onChange={onChange}
        />
        <label className="custom-control-label" htmlFor="customCheck">
          {labeling}
        </label>
      </div>
    );
  }
}

/**
 * Renders a form select using Bootstrap styles.
 */
class FormSelect extends Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLSelectElement>) => mixed,
  children?: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, size.
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, children, ...rest } = this.props;
    return (
      <select {...rest} className="custom-select" value={value} onChange={onChange}>
        {children}
      </select>
    );
  }
}

/**
 * Renders a form date using Bootstrap styles.
 */
class FormDate extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders form components using Bootstrap styles.
 */
export default class Form {
  static Label = FormLabel;
  static Input = FormInput;
  static Textarea = FormTextarea;
  static Checkbox = FormCheckbox;
  static Select = FormSelect;
  static Date = FormDate;
}
