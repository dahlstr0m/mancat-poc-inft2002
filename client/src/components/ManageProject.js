// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { history } from '../index';
import Button from './Button';
import Form from './Form';
import { Card } from './Card';
import { Alert } from './Widgets';
import { projectService, categoryService, employerService } from '../services';

/**
 * Renders page to Mange Project
 */
export default class ManageProject extends Component<{ pathId: number }> {
  project: Project = {
    // Default values for project, updated on change in form
    projectId: 0,
    title: '',
    projectDescription: '',
    projectDate: '',
    categoryId: 1,
    employerId: 1,
    ranking: 0,
    active: false,
  };
  categories: Category[] = [];
  employers: Employer[] = [];
  render() {
    return (
      <>
        <p>
          Verdier
          <br />
          projectId: {this.project.projectId} <br />
          title: {this.project.title} <br />
          description: {this.project.projectDescription} <br />
          date: {this.project.projectDate} <br />
          category: {this.project.categoryId} <br />
          employer: {this.project.employerId} <br />
          active?:{' '}
          {this.project.active == true
            ? 'Aktivt'
            : this.project.active == false
            ? 'Deaktivert'
            : 'error'}{' '}
          <br />
        </p>
        <Card title={'Project: ' + this.project.title}>
          <Form.Label> Title:</Form.Label>
          <Form.Input
            type="text"
            value={this.project.title}
            onChange={(event) => (this.project.title = event.currentTarget.value)}
          ></Form.Input>
          <br />
          <Form.Label> Description:</Form.Label>
          <Form.Textarea
            value={this.project.projectDescription}
            onChange={(event) => (this.project.projectDescription = event.currentTarget.value)}
          ></Form.Textarea>
          <br />
          <Form.Label> Date:</Form.Label>
          <Form.Input
            type="date"
            value={this.project.projectDate.substring(0, 10)}
            onChange={(event) => (this.project.projectDate = event.currentTarget.value)}
          ></Form.Input>
          <br />
          <Form.Label> Category:</Form.Label>
          <Form.Select
            value={this.project.categoryId}
            onChange={(event) => (this.project.categoryId = parseInt(event.currentTarget.value))}
          >
            {this.categories.map((category) => (
              <option key={category.categoryId} value={category.categoryId}>
                {category.categoryName}
              </option>
            ))}
          </Form.Select>
          <br />
          <Form.Label> Employer:</Form.Label>
          <Form.Select
            value={this.project.employerId}
            onChange={(event) => (this.project.employerId = parseInt(event.currentTarget.value))}
          >
            {this.employers.map((employer) => (
              <option key={employer.employerId} value={employer.employerId}>
                {employer.employerName}
              </option>
            ))}
          </Form.Select>
          <br />
          <br />
          <Form.Checkbox
            checked={this.project.active}
            onChange={(event) => (this.project.active = event.currentTarget.checked)}
            labeling={'Do you want the project active from start?'}
          ></Form.Checkbox>
          <br />
          <Button.Success
            onClick={() =>
              projectService
                .updateProject(this.project)
                .then(() => {
                  history.push('/admin');
                  Alert.success('Project successfully saved');
                })
                .catch((error: Error) => Alert.danger('Error storing new project: ' + error))
            }
          >
            Save project
          </Button.Success>
          <Button.Danger
            onClick={() =>
              projectService
                .deleteProject(this.project.projectId)
                .then(() => {
                  history.push('/admin');
                  Alert.success('Project successfully deleted');
                })
                .catch((error: Error) => Alert.danger('Error deleting project: ' + error))
            }
          >
            Delete project
          </Button.Danger>
        </Card>
      </>
    );
  }
  mounted() {
    projectService
      .getProject(this.props.pathId)
      .then((project) => (this.project = project))
      .then(() => (this.project.projectDate = this.project.projectDate.substring(0, 10)))
      .catch((error: Error) => Alert.danger('Error getting project: ' + error));
    categoryService
      .getCategories()
      .then((categories) => (this.categories = categories))
      .catch((error: Error) => Alert.danger('Error getting categories: ' + error));

    employerService
      .getEmployers()
      .then((employers) => (this.employers = employers))
      .catch((error: Error) => Alert.danger('Error getting employers: ' + error));
  }
}
