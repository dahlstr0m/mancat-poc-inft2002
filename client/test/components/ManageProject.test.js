// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

// Components
import ManageProject from '../../src/components/ManageProject';
import Form from '../../src/components/Form';
import Button from '../../src/components/Button';

// Mocks
jest.mock('../../src/services/project-service');
jest.mock('../../src/services/poster-service');
jest.mock('../../src/services/category-service');
jest.mock('../../src/services/employer-service');

// Tests
describe('ManageProject component tests', () => {
  test('ManageProject draws correctly', (done) => {
    const wrapper = shallow(<ManageProject pathId={1} />);

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();

      done();
    });
  });

  test('ManageProject saves project correctly', (done) => {
    const wrapper = shallow(<ManageProject pathId={1} />);

    wrapper
      .find(Form.Input)
      .at(0)
      .simulate('change', { currentTarget: { value: 'New title' } });
    wrapper.find(Form.Textarea).simulate('change', { currentTarget: { value: 'New desc' } });
    wrapper
      .find(Form.Input)
      .at(1)
      .simulate('change', { currentTarget: { value: 'New date' } });
    wrapper
      .find(Form.Select)
      .at(0)
      .simulate('change', { currentTarget: { value: 2 } });
    wrapper
      .find(Form.Select)
      .at(1)
      .simulate('change', { currentTarget: { value: 2 } });
    wrapper.find(Form.Checkbox).simulate('change', { currentTarget: { checked: true } });

    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input value="New title" />)).toEqual(true);
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Textarea value="New desc" />)).toEqual(true);
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Input value="New date" />)).toEqual(true);
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Select value={2} />)).toEqual(true);
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Select value={2} />)).toEqual(true);
    // $FlowExpectedError
    expect(wrapper.containsMatchingElement(<Form.Checkbox checked={true} />)).toEqual(true);

    wrapper.find(Button.Success).simulate('click');

    setTimeout(() => {
      expect(location.hash).toEqual('#/admin');
      done();
    });
  });

  test('ManageProject removes project correctly', (done) => {
    const wrapper = shallow(<ManageProject pathId={1} />);

    wrapper.find(Button.Danger).simulate('click');

    setTimeout(() => {
      expect(location.hash).toEqual('#/admin');
      done();
    });
  });
});
