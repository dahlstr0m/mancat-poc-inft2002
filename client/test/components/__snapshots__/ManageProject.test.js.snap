// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageProject component tests ManageProject draws correctly 1`] = `
<Fragment>
  <Card
    title="Project: Project 1"
  >
    <Card>
      <p>
        No posters in project
      </p>
    </Card>
  </Card>
  <Card
    title="Project: Project 1"
  >
    <FormLabel>
       Title:
    </FormLabel>
    <FormInput
      onChange={[Function]}
      type="text"
      value="Project 1"
    />
    <br />
    <FormLabel>
       Description:
    </FormLabel>
    <FormTextarea
      onChange={[Function]}
      value="Project 1 description"
    />
    <br />
    <FormLabel>
       Date:
    </FormLabel>
    <FormInput
      onChange={[Function]}
      type="date"
      value="10-10-2020"
    />
    <br />
    <FormLabel>
       Category:
    </FormLabel>
    <FormSelect
      onChange={[Function]}
      value={1}
    >
      <option
        key="1"
        value={1}
      >
        Kategori 1
      </option>
      <option
        key="2"
        value={2}
      >
        Kategori 2
      </option>
      <option
        key="3"
        value={3}
      >
        Kategori 3
      </option>
      <option
        key="4"
        value={4}
      >
        Kategori 4
      </option>
    </FormSelect>
    <br />
    <FormLabel>
       Employer:
    </FormLabel>
    <FormSelect
      onChange={[Function]}
      value={1}
    >
      <option
        key="1"
        value={1}
      >
        Employer 1
      </option>
      <option
        key="2"
        value={2}
      >
        Employer 2
      </option>
      <option
        key="3"
        value={3}
      >
        Employer 3
      </option>
      <option
        key="4"
        value={4}
      >
        Employer 4
      </option>
    </FormSelect>
    <br />
    <br />
    <FormCheckbox
      checked={true}
      labeling="Do you want the project active?"
      onChange={[Function]}
    />
    <br />
    <ButtonSuccess
      onClick={[Function]}
    >
      Save project
    </ButtonSuccess>
    <ButtonLight
      onClick={[Function]}
    >
      Delete project
    </ButtonLight>
    <div
      className="d-none"
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Posters and project will be deleted
      </ButtonDanger>
    </div>
  </Card>
</Fragment>
`;
